File: auxiliary/Thrift/src/greta/auxiliary/thrift/services/Connector.java
/*
 * Copyright 2025 Greta Modernization Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package greta.auxiliary.thrift.services;

import greta.core.util.log.Logs;

/*
 * Sender imports
 *
 * import greta.core.util.log.Logs; import org.apache.thrift.TException; import
 * org.apache.thrift.protocol.TBinaryProtocol; import
 * org.apache.thrift.protocol.TProtocol; import
 * org.apache.thrift.transport.TSocket; import
 * org.apache.thrift.transport.TTransport;
 *
 */

/*
 * Receiver imports * import org.apache.thrift.server.TServer; import
 * org.apache.thrift.server.TServer.Args; import
 * org.apache.thrift.server.TSimpleServer; import
 * org.apache.thrift.transport.TServerSocket; import
 * org.apache.thrift.transport.TServerTransport;
 *
 */
/**
 *
 * @author Ken Prepin
 */
public abstract class Connector<C extends Connector> {

    public static final String DEFAULT_THRIFT_HOST = "localhost";
    public static final int DEFAULT_THRIFT_PORT = 9090;
    public static final int SLEEP_TIME = 700;
    private String host;
    private int port;
    private ConnectionListener connectionListener;
    ConnectionStarter starter;

    /*
     * = isStarted for Servers
     */
    private Boolean isConnected;

    public Connector() {
        this(DEFAULT_THRIFT_HOST, DEFAULT_THRIFT_PORT);
    }

    public Connector(String host, int port) {
        this.host = host;
        this.port = port;
        this.isConnected = false;
        this.startConnection();
    }

    @Override
    protected void finalize() throws Throwable {
        starter = null;
        super.finalize();
    }

    public String getHost() {
        return host;
    }

    public String getPortString() {
        return String.valueOf(port);
    }

    public int getPort() {
        return port;
    }

    public void setHost(String host) {
        this.host = host;
        startConnection();
    }

    public void setPort(int port) {
        this.port = port;
        startConnection();
    }

    public void setPort(String port) {
        this.port = Integer.parseInt(port);
        startConnection();
    }

    public void setURL(String host, String port) {
        setHost(host);
        setPort(port);
        startConnection();
    }


    public void setConnected(boolean isConnected) {
        this.isConnected = isConnected;
        if (connectionListener != null) {
            if (isConnected) {
                connectionListener.onConnection();
            } else {
                connectionListener.onDisconnection();
            }
        }
    }

    public void startConnection() {
        if (this.isConnected()) {
            this.starter.stopConnector(this);
        }
        setConnected(false);
        starter = new ConnectionStarter(this);
        starter.start();

    }

    private static class ConnectionStarter extends Thread {

        final Connector connector;

        public ConnectionStarter(Connector connectorToBeStarted) {
            this.setDaemon(true);
            this.connector = connectorToBeStarted;
        }

        @Override
        public void run() {
            Logs.debug("Try to start on " + connector.getHost() + " - " + connector.getPort());
            while (!connector.isConnected() && connector.starter == this) {
                connector.startConnector(connector);
                if (!connector.isConnected()) {
                    try {
                        sleep(SLEEP_TIME);
                    } catch (Exception ex1) {
                    }
                }
            }
        }

         public void stopConnector(Connector connector){
             this.connector.stopConnector(connector);
         };
    }

    public abstract void startConnector(C connector);

    public abstract void stopConnector(C connector);

    public boolean isConnected() {
        return isConnected;
    }

    public void setConnectionListener(ConnectionListener connectionListener) {
        this.connectionListener = connectionListener;
    }
}

File: gretaUnity-master-1/gretaUnity-master/Assets/Scripts/ThriftImpl/services/Connector.cs
using System;
using UnityEngine;
using System.Threading;

/**
 * @author: Ken
 *
 */

namespace thrift.services
{
	public abstract class Connector
	{
		
		public static String DEFAULT_THRIFT_HOST = "localhost";
		public static int DEFAULT_THRIFT_PORT = 9095;
		public static int SLEEP_TIME = 1000;
		private String host;
		private int port;
		private ConnectionListener connectionListener;
		Thread starterTh;
		private bool onConnection;
		/*
     * = isStarted for Servers
     */
		private bool connected;

		public Connector () : this(DEFAULT_THRIFT_HOST, DEFAULT_THRIFT_PORT)
		{
		}

		public Connector (String host, int port)
		{
			this.host = host;
			this.port = port;
			this.connected = false;
			this.onConnection=false;
		}

		~Connector ()
		{
			stopConnector ();			
		}

		public String getHost ()
		{
			return host;
		}

		public String getPortString ()
		{
			return Convert.ToString (port);
		}

		public int getPort ()
		{
			return port;
		}

		public void setHost (String host)
		{
			this.host = host;
			stopConnector ();
			//startConnection ();
		}

		public void setPort (int port)
		{
			this.port = port;
			stopConnector ();
			//startConnection ();
		}

		public void setPort (String port)
		{
			this.port = Convert.ToInt16 (port);
			stopConnector ();
			//startConnection ();
		}

		public void setURL (String host, String port)
		{
			setHost (host);
			setPort (port);
			stopConnector ();
			//startConnection ();
		}

		public void setURL (String host, int port)
		{
			setHost (host);
			setPort (port);
			stopConnector ();
			//startConnection ();
		}

		public void setConnectionLisnter (ConnectionListener connectionListener)
		{
			this.connectionListener = connectionListener;
		}

		public void setConnected (bool isConnected)
		{
			this.connected = isConnected;
			if (connectionListener != null) {
				if (connected) {
					connectionListener.onConnection ();
				} else {
					connectionListener.onDisconnection ();
				}
			}
		}

		public void startConnection ()
		{
			//Debug.Log ("in StartConnection");
			if (this.isConnected ()) {
				setConnected (false);
				this.stopConnector ();
			}  
			onConnection = true;
			starterTh = new Thread (new ThreadStart (connectionStarting));
			starterTh.IsBackground = true;  
			starterTh.Start ();
		}

		public void connectionStarting ()
		{
			
			int cpt = 1;
			while (!this.isConnected()&&Thread.CurrentThread == starterTh) {
				//Debug.Log ("Try to start connection on " + this.getHost () + " - " + this.getPort () + " " + cpt);
				cpt++;
				this.startConnector ();
				if (!this.isConnected ()) {
					try {
						Thread.Sleep (SLEEP_TIME);
					} catch (Exception ex1) {
						Debug.LogError (ex1);
					}
				} else {
					onConnection = false;
				}
			}
		}

		public abstract void startConnector ();

		public abstract void stopConnector ();

		public bool isConnected ()
		{
			return connected;
		}
		public bool isOnConnection(){
			return onConnection;
		}

		public void setConnectionListener (ConnectionListener connectionListener)
		{
			this.connectionListener = connectionListener;
		}
	}
}

